lib/experimental/interactivity-api/directive-processing.php

```diff
@@ -0,0 +1,156 @@
+		return $block_content;
+	}
+
+	// TODO: Add some directive/components registration mechanism.
+	$directives = array(
+		'data-wp-bind'    => 'gutenberg_interactivity_process_wp_bind',
+		'data-wp-context' => 'gutenberg_interactivity_process_wp_context',
+		'data-wp-class'   => 'gutenberg_interactivity_process_wp_class',
+		'data-wp-style'   => 'gutenberg_interactivity_process_wp_style',
+		'data-wp-text'    => 'gutenberg_interactivity_process_wp_text',
+	);
+
+	$tags = new WP_Directive_Processor( $block_content );
+	$tags = gutenberg_interactivity_process_directives( $tags, 'data-wp-', $directives );
+	return $tags->get_updated_html();
+}
+add_filter( 'render_block', 'gutenberg_interactivity_process_directives_in_root_blocks', 10, 2 );
+
+/**
+ * Mark the inner blocks with a temporary property so we can discard them later,
+ * and process only the root blocks.
+ *
+ * @param array $parsed_block The parsed block.
+ * @param array $source_block The source block.
+ * @param array $parent_block The parent block.
+ *
+ * @return array The parsed block.
+ */
+function gutenberg_interactivity_mark_inner_blocks( $parsed_block, $source_block, $parent_block ) {
+	if ( isset( $parent_block ) ) {
+		$parsed_block['is_inner_block'] = true;
+	}
+	return $parsed_block;
+}
+add_filter( 'render_block_data', 'gutenberg_interactivity_mark_inner_blocks', 10, 3 );
+
+/**
+ * Process directives.
+ *
+ * @param WP_Directive_Processor $tags An instance of the WP_Directive_Processor.
+ * @param string                 $prefix Attribute prefix.
+ * @param string[]               $directives Directives.
+ *
+ * @return WP_Directive_Processor The modified instance of the
+ * WP_Directive_Processor.
+ */
+function gutenberg_interactivity_process_directives( $tags, $prefix, $directives ) {
+	$context   = new WP_Directive_Context;
+	$tag_stack = array();
+
+	while ( $tags->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+		$tag_name = strtolower( $tags->get_tag() );
+
+		// Is this a tag that closes the latest opening tag?
+		if ( $tags->is_tag_closer() ) {
+			if ( 0 === count( $tag_stack ) ) {
+				continue;
+			}
+
+			list( $latest_opening_tag_name, $attributes ) = end( $tag_stack );
+			if ( $latest_opening_tag_name === $tag_name ) {
+				array_pop( $tag_stack );
+
+				// If the matching opening tag didn't have any attribute directives,
+				// we move on.
+				if ( 0 === count( $attributes ) ) {
+					continue;
+				}
+			}
+		} else {
+			// Helper that removes the part after the double hyphen before looking for
+			// the directive processor inside `$attribute_directives`.
+			$get_directive_type = function ( $attr ) {
+				return explode( '--', $attr )[0];
+			};
+
+			$attributes = $tags->get_attribute_names_with_prefix( $prefix );
+			$attributes = array_map( $get_directive_type, $attributes );
+			$attributes = array_intersect( $attributes, array_keys( $directives ) );
```

Comment:
for cases like this in performance sensitive code (of which this very much is) it would be good to adopt a default practice of avoiding multiple iterations and array functions. we can avoid a function call and avoid the costly array iterations and intersection with a direct `foreach` loop

```php
$attributes = array();
foreach ( $tags->get_attribute_names_with_prefix( $prefix ) as $name ) {
	// Get "wp-bind" from "wp-bind--src" and "wp-context" from "wp-context" etc...
	list( $type ) = explode( '--', $name );
	if ( array_key_exists( $type, $directives ) ) {
		$attributes[] = $name;
	}
}
```